### PR TITLE
Plane: disable pitch based weathervaning for quadplanes

### PR DESCRIPTION
**Pull request by tridge**
planes with a pitch asymmetry are very common, as used move payload slightly or have different front motors from rear motors (common in tilt quadplanes)

constantly yawing in no wind is annoying, so best to just disable pitch for weathervaning for nose-in and tail-in.

### CODE REVIEW

**Review by IamPete1 — CHANGES_REQUESTED** [review]
This change brakes side in weathervaning.

Including pitch is a big advantage for nose in weathervaning when you happen to be 180 deg wrong. In that case your roll is also very close to zero. 

We actually do not apply the deadzone in the positive pitch case. 

https://github.com/ArduPilot/ardupilot/blob/1240ed13b1dbc85a8a496929bfe3c8c00469b0af/libraries/AC_AttitudeControl/AC_WeatherVane.cpp#L168-L170 

The assumption was that you must be nose down to hold into the wind. 

Maybe that should be the first step?
```
            if (is_positive(pitch_cdeg-deadzone_cdeg)) {
                output = fabsf(roll_cdeg) + pitch_cdeg - deadzone_cdeg;
            } else {
```

**Comment by Gone4Dirt** [comment]
yeah, please please don't do this.  Adding a deadzone should solve this.

**Comment by tridge** [comment]
> yeah, please please don't do this. Adding a deadzone should solve this.

I don't think a deadzone helps. It is quite common for quadplanes to hover in zero wind with either significant nose-up or nose-down. You'd need a huge deadzone, otherwise existing aircraft will start spinning in zero wind, which is exactly what we have seen happen.
I've added an OPTIONS parameter to control use of pitch for nose-in and tail-in

**Comment by tridge** [comment]
i've added in the pitch dead-zone as well, but we still need the options. Hovering with 5 degrees pitch up isn't uncommon in some quadplanes. I don't want those to start spinning in 4.2 release.

File: libraries/AC_AttitudeControl/AC_WeatherVane.cpp [hunk 1]
```diff
@@ -159,13 +166,17 @@ bool AC_WeatherVane::get_yaw_out(float &yaw_output, const int16_t pilot_yaw, con
...
+
     switch (dir) {
         case Direction::OFF:
             reset();
             return false;
 
         case Direction::NOSE_IN:
-            if (is_positive(pitch_cdeg)) {
+            if (pitch_enable && is_positive(pitch_cdeg - deadzone_cdeg)) {
                 output = fabsf(roll_cdeg) + pitch_cdeg;
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_WeatherVane.cpp:180]:
```suggestion
                output = fabsf(roll_cdeg) + pitch_cdeg - deadzone_cdeg;
```
Needs to match the `is_positive` amount so there is no step change when you move from using pitch to not.

**Review by IamPete1 — COMMENTED** [review]
How about using double the set deadzone for pitch? Such issues are exactly what the deadzone param is for, a vehicle with a roll offset will also yaw for ever. 

If planes are hovering with 5deg pitch up they should really use Q_TRIM_PITCH to make that 0.

**Comment by Hwurzburg** [comment]
but some nose up hover is actually desirable to reduce the required hover lift  in a headwind and VTOL horizontal movement....

File: libraries/AC_AttitudeControl/AC_WeatherVane.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AC_AttitudeControl/AC_WeatherVane.cpp]: thanks, fixed (in 2 places)

**Comment by tridge** [comment]
> How about using double the set deadzone for pitch? Such issues are exactly what the deadzone param is for, a vehicle with a roll offset will also yaw for ever.

two problems with that:
 - roll asymmetry is quite rare, whereas pitch asymmetry is common
 - we are updating existing aircraft, and users expect their existing setup to keep working on update to 4.2.x. If they had a roll asymmetry they would already have changed their deadzone to cope or disabled weathervaning

> If planes are hovering with 5deg pitch up they should really use Q_TRIM_PITCH to make that 0.

they could, but we should not break existing aircraft with this change to weathervaning
plus the pitch trim often changes between flights based on battery placement and payload

File: libraries/AC_AttitudeControl/AC_WeatherVane.cpp [hunk 2]
```diff
@@ -97,7 +104,7 @@ AC_WeatherVane::AC_WeatherVane(void)
     AP_Param::setup_object_defaults(this, var_info);
 }
 
-bool AC_WeatherVane::get_yaw_out(float &yaw_output, const int16_t pilot_yaw, const float hgt, const float roll_cdeg, const float pitch_cdeg, const bool is_takeoff, const bool is_landing)
+bool AC_WeatherVane::get_yaw_out(float &yaw_output, const int16_t pilot_yaw, const float hgt, const float roll_cdeg, float pitch_cdeg, const bool is_takeoff, const bool is_landing)
```
Comment (IamPete1) [inline libraries/AC_AttitudeControl/AC_WeatherVane.cpp:107]: Why have we lost the const?

Comment (tridge) [reply libraries/AC_AttitudeControl/AC_WeatherVane.cpp]: fixed

**Review by Gone4Dirt — APPROVED** [review]

**Comment by Hwurzburg** [comment]
@IamPete1 so I am trying to write this up.....does this change current behavior or not (just for my info)?

**Comment by IamPete1** [comment]
@Hwurzburg No change vs 4.1. Including pitch was in master for abit but now is a option only.

**Pull request merged**